plugins/authentication/src/ExternalTokenModel/model.tsx

```diff
@@ -0,0 +1,150 @@
+          (location as AuthLocation)?.uri.includes(domain),
+        )
+      },
+    }))
+    .actions(self => ({
+      async fetchFile(location: string) {
+        if (!location || !self.externalToken) {
+          return
+        }
+        // add a fetch call for gdc adding the token to the header, or place the header into
+        // sessionstorage for fetch to use
+      },
+      setExternalToken(token: string) {
+        self.externalToken = token
+      },
+      setNeedsToken(bool: boolean) {
+        self.needsToken = bool
+      },
+      getOrSetExternalToken() {
+        if (!self.needsToken) {
+          return ''
+        }
+        const tokenKey = Object.keys(sessionStorage).find(key => {
+          return key === `${self.accountConfig.internetAccountId}-token`
+        })
+        let token = sessionStorage.getItem(tokenKey as string)
+
+        // prompt user for token if there isnt one existing
+        // if a user doesnt enter a token allow them to continue without token
+
+        // for GDC there is a way to tell if it needs a token or not
+        // first call is an api endpoint, you get metadata for the file and has a property
+        // called access, if access is set to controller it needs a token, else it doesn't
+        if (!token) {
+          const newToken = window.prompt(
+            `Enter token for ${self.accountConfig.name} to use`,
+          )
+          if (!newToken) {
+            return
+          }
+          token = newToken
+          sessionStorage.setItem(
+            `${self.accountConfig.internetAccountId}-token`,
+            token,
+          )
+        }
+
+        return token
+      },
+      async openLocation(location: FileLocation) {
+        switch (self.accountConfig.origin) {
+          case 'GDC': {
+            const query = (location as AuthLocation).uri.split('/').pop() // should get id
+            const response = await fetch(
+              `${self.accountConfig.customEndpoint}/files/${query}?expand=index_files`,
+              {
+                method: 'GET',
+              },
+            )
+
+            if (!response.ok) {
+              const errorText = await response.text()
+              throw new Error(
+                `Network response failure: ${response.status} (${errorText})`,
+              )
+            }
+
+            const metadata = await response.json()
+            if (metadata) {
+              metadata.access === 'controlled'
+                ? this.setNeedsToken(true)
+                : this.setNeedsToken(false)
+            }
+
+            this.getOrSetExternalToken()
+          }
+        }
+      },
+      handleRpcMethodCall(
+        location: FileLocation,
+        authenticationInfoMap: Record<string, string>,
+        args: {},
+      ) {
+        const token =
+          authenticationInfoMap[self.accountConfig.internetAccoundId]
+        if (!token) {
+          this.openLocation(location)
```

Comment:
this needs to be (i think) `await this.openLocation(location)` (otherwise the next code block will execute without the token and the initial call will fail with a 403)